### PR TITLE
Media player rendering with GL textures

### LINKED ISSUE #22300: Enable hardware accelerated video playback
This involves several pieces, some of them are already WIP:

- [x] [Add bindings for GStreamerGL](https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/merge_requests/161)
- [x] [Expose the EGLDisplay](https://github.com/tomaka/glutin/issues/1078)
- [x] [Modify player example to use GStreamerGL](https://github.com/servo/media/issues/166)
- [x] [Add a way to share EGLDisplay/X11 Display and GL context with servo-media](https://github.com/servo/servo/issues/22920)
- [x] Modify the existing HTMLMediaElement FrameRenderer to use GStreamerGL

### LINKED ISSUE #22920: Use new GL-based video playback
In answer to @ceyusa in https://mozilla.logbot.info/servo/20190221#c15994865-c15994866:
* obtain the relevant glutin data inside ports/servo/glutin_app
* add an API to [WindowMethods](https://github.com/servo/servo/blob/0dda1156094dae5ff47ab6fcf316cdc5547fff4c/components/compositing/windowing.rs#L130) to receive that data
* add the required data to [InitialPipelineState](https://github.com/servo/servo/blob/0dda1156094dae5ff47ab6fcf316cdc5547fff4c/components/constellation/pipeline.rs#L98)
* add the required data to Constellation so it can be added to the pipeline state
* add the required data to ScriptThreadFactory's [InitialScriptState](https://github.com/servo/servo/blob/0dda1156094dae5ff47ab6fcf316cdc5547fff4c/components/script_traits/lib.rs#L548)
* add the data to Window, so it can be obtained by HTMLMediaElement when necessary
* store the data in script's ScriptThread so it can be passed to the Window constructor

### PR DESCRIPTION
**Pull request by ceyusa**
These patches pass the application's OpenGL raw context  and the its native display address to the media player, in order to create an internal wrapped context, thus it will generate video frames as textures.

For now only EGL from glutin-based app and android are in place, though tested only in Linux glutin app.

This PR also renders the generated frame textures by Servo/Media and renders them by using a thread that connects Webrenderer's ExternalImageHandler and each instantiated player. **By now, these patches, disable the WebGL rendering**. We need to provide a ExternalImageHandler demuxer.

This PR depends on https://github.com/servo/media/pull/270

- [X]  `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- This PR fixes #22300 and fixes #22920

In order to test it you must launch servo as

`./mach run -- --pref media.glvideo.enabled [...]`

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23483)

### CODE REVIEW

**Comment by jdm** [comment]
Is this waiting for review right now?

**Comment by ceyusa** [comment]
> Is this waiting for review right now?

Is still a working progress :( 

It would be great a glance to check if the sketched ideas make senss.

**Comment by ceyusa** [comment]
Ready for review :)

File: components/script/dom/htmlmediaelement.rs [hunk 1]
```diff
@@ -83,11 +83,12 @@ use std::mem;
 use std::rc::Rc;
 use std::sync::{Arc, Mutex};
 use time::{self, Duration, Timespec};
-//use webrender_api::{ExternalImageData, ExternalImageId, ExternalImageType, TextureTarget};
+use webrender_api::{ExternalImageData, ExternalImageId, ExternalImageType, TextureTarget};
 use webrender_api::{ImageData, ImageDescriptor, ImageFormat, ImageKey, RenderApi};
 use webrender_api::{RenderApiSender, Transaction};
 
 pub struct MediaFrameRenderer {
+    id: u64,
```
Comment (jdm) [inline components/script/dom/htmlmediaelement.rs]: Let's use an Option and a more descriptive name. player_id?

File: components/script/dom/htmlmediaelement.rs [hunk 2]
```diff
@@ -156,17 +158,17 @@ impl FrameRenderer for MediaFrameRenderer {
...
-                    //     descriptor,
-                    //     ImageData::External(ExternalImageData {
-                    //         id: ExternalImageId(0), // let's try to fool webgl
-                    //         channel_index: 0,
-                    //         image_type: ExternalImageType::TextureHandle(TextureTarget::Default),
-                    //     }),
-                    //     None,
-                    // );
+                } else if self.id != 0 {
+                    txn.add_image(
```
Comment (jdm) [inline components/script/dom/htmlmediaelement.rs]: Let's merge these blocks:
```rust
let image_data = if let Some(id) = self.player_id {
   ImageData::External(...)
} else {
   ImageData::Raw(frame.get_data())
};
txn.add_image(new_image_key, descriptor, image_data, None);
```

File: components/script/dom/htmlmediaelement.rs [hunk 3]
```diff
@@ -184,17 +186,17 @@ impl FrameRenderer for MediaFrameRenderer {
...
-                    //     image_key,
-                    //     descriptor,
-                    //     ImageData::External(ExternalImageData {
-                    //         id: ExternalImageId(0), // let's try to fool webgl
-                    //         channel_index: 0,
-                    //         image_type: ExternalImageType::TextureHandle(TextureTarget::Default),
-                    //     }),
-                    //     None,
-                    // );
+                } else if self.id != 0 {
```
Comment (jdm) [inline components/script/dom/htmlmediaelement.rs]: Some comments as previous case - let's merge these blocks.

**Review by jdm — CHANGES_REQUESTED** [review]
Exciting! This looks like a sensible implementation.

File: ports/glutin/headless_window.rs [hunk 4]
```diff
@@ -195,4 +196,16 @@ impl WindowMethods for Window {
     }
 
     fn prepare_for_composite(&self) { }
+
+    fn get_gl_context(&self) -> MediaPlayerCtxt::GlContext {
```
[the remark this answers is not in the document]
Comment (jdm) [reply ports/glutin/headless_window.rs]: The point of this comment is to spur conversation - it feels like something we _should_ test if possible; how much work is it to support OSMesa for the GL video rendering integration?

**Comment by jdm** [comment]
Are the webrender external image changes ready for review? I saw some comments that suggested that some more work is expected.

**Comment by ferjm** [comment]
Not ready for review yet. I need to add a few more changes.

File: components/script/dom/window.rs [hunk 5]
```diff
@@ -300,6 +301,10 @@ pub struct Window {
     /// Flag that indicates if the layout thread is busy handling a request.
     #[ignore_malloc_size_of = "Arc<T> is hard"]
     layout_is_busy: Arc<AtomicBool>,
+
+    /// Window's GL context from application
+    #[ignore_malloc_size_of = "defined in script_thread"]
```
[the remark this answers is not in the document]
Comment (ferjm) [reply components/script/dom/window.rs]: `WindowGLContext` has `GlContext`, `GlApi` and `NativeDisplay` members that are defined in servo-media. The 4th member is a `GLPlayerPipeline`, which is the only one defined in Servo, but it is basically an `IpcSender` or a `mpsc::Sender`, also defined outside of Servo.

File: components/webrender_traits/lib.rs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            let (texture_id, size) = match handler_type {
+                WebrenderImageHandlerType::WebGL => {
+                    self.webgl_handler.as_mut().unwrap().lock(key.0)
+                },
+                WebrenderImageHandlerType::Media => {
+                    self.media_handler.as_mut().unwrap().lock(key.0)
+                },
+            };
+            webrender::ExternalImage {
+                uv: webrender_api::TexelRect::new(0.0, 0.0, size.width as f32, size.height as f32),
```
Comment (ceyusa) [inline components/webrender_traits/lib.rs]: Sadly, textures from media and textures from WebGL don't have the same coordinates in texel space:

webgl: 0.0, size.height as f32, size.width as f32, 0.0

media (as gstreamer): 0.0, 0.0, size.width as f32, size.height as f32

File: ports/glutin/headless_window.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ceyusa) [reply ports/glutin/headless_window.rs]: I have never tried GstGL on OSMesa. So far I have no idea.

File: components/media/lib.rs [hunk 7]
```diff
@@ -2,12 +2,99 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
 
-use crate::media_thread::{GLPlayerExternalImageApi, GLPlayerExternalImageHandler, GLPlayerThread};
-use canvas_traits::media::glplayer_channel;
-use canvas_traits::media::{
-    GLPlayerChan, GLPlayerMsg, GLPlayerPipeline, GLPlayerReceiver, GLPlayerSender,
-};
+#![crate_name = "media"]
+#![crate_type = "rlib"]
```
Comment (jdm) [inline components/media/lib.rs]: These crate attributes shouldn't be necessary.

File: components/webrender_traits/lib.rs [hunk 8]
```diff
@@ -0,0 +1,108 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#![crate_name = "webrender_traits"]
+#![crate_type = "rlib"]
```
Comment (jdm) [inline components/webrender_traits/lib.rs]: These crate attributes shouldn't be necessary.

File: components/webrender_traits/lib.rs [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+            }
+        } else {
+            unreachable!()
+        }
+    }
+
+    /// Unlock the external image. The WR should not read the image
+    /// content after this call.
+    fn unlock(&mut self, key: webrender_api::ExternalImageId, _channel_index: u8) {
+        if let Some(handler_type) = self.external_images.get(&key) {
```
Comment (jdm) [inline components/webrender_traits/lib.rs]: Let's use `expect` here instead.

File: components/webrender_traits/lib.rs [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    /// image content.
+    /// The WR client should not change the image content until the
+    /// unlock() call.
+    fn lock(
+        &mut self,
+        key: webrender_api::ExternalImageId,
+        _channel_index: u8,
+        _rendering: webrender_api::ImageRendering,
+    ) -> webrender::ExternalImage {
+        if let Some(handler_type) = self.external_images.get(&key) {
```
Comment (jdm) [inline components/webrender_traits/lib.rs]: Let's use `expect` here instead.

**Review by jdm — APPROVED** [review]
Looks good!

**Comment by jdm** [comment]
Building simpleservo:
```
[0m[1m[38;5;9merror[E0063][0m[0m[1m: missing fields `gl_context_pointer`, `native_display_pointer` in initializer of `simpleservo::InitOptions`[0m
[0m   [0m[0m[1m[38;5;12m--> [0m[0mports/libsimpleservo/capi/src/lib.rs:120:16[0m
[0m    [0m[0m[1m[38;5;12m|[0m
[0m[1m[38;5;12m120[0m[0m [0m[0m[1m[38;5;12m| [0m[0m    let opts = InitOptions {[0m
[0m    [0m[0m[1m[38;5;12m| [0m[0m               [0m[0m[1m[38;5;9m^^^^^^^^^^^[0m[0m [0m[0m[1m[38;5;9mmissing `gl_context_pointer`, `native_display_pointer`[0m

[0m[1m[38;5;9merror[0m[0m[1m: aborting due to previous error[0m
```

Building windows:
```
error: unused import: `servo::gl::GlesFns`
  --> ports\libsimpleservo\api\src\gl_glue.rs:12:9
   |
12 |     use servo::gl::GlesFns;
   |         ^^^^^^^^^^^^^^^^^^
   |
   = note: `-D unused-imports` implied by `-D warnings`

error: unused import: `std::ffi::CString`
  --> ports\libsimpleservo\api\src\gl_glue.rs:13:9
   |
13 |     use std::ffi::CString;
   |         ^^^^^^^^^^^^^^^^^

error: unused import: `std::os::raw::c_void`
  --> ports\libsimpleservo\api\src\gl_glue.rs:14:9
   |
14 |     use std::os::raw::c_void;
   |         ^^^^^^^^^^^^^^^^^^^^

error: aborting due to 3 previous errors

error: Could not compile `simpleservo`.
```

Building magic leap:
```
error[E0599]: no method named `clone` found for type `simpleservo::gl_glue::egl::EGLInitResult` in the current scope
   --> ports/libmlservo/src/lib.rs:148:52
    |
148 |             MagicLeapVRService::new(name, ctxt, gl.clone()).expect("Failed to create VR service");
    |                                                    ^^^^^

error[E0063]: missing fields `gl_context_pointer`, `native_display_pointer` in initializer of `simpleservo::InitOptions`
   --> ports/libmlservo/src/lib.rs:153:16
    |
153 |     let opts = InitOptions {
    |                ^^^^^^^^^^^ missing `gl_context_pointer`, `native_display_pointer`

error[E0308]: mismatched types
   --> ports/libmlservo/src/lib.rs:175:29
    |
175 |     simpleservo::init(opts, gl, wakeup, callbacks).expect("error initializing Servo");
    |                             ^^ expected struct `std::rc::Rc`, found struct `simpleservo::gl_glue::egl::EGLInitResult`
    |
    = note: expected type `std::rc::Rc<(dyn gleam::gl::Gl + 'static)>`
               found type `simpleservo::gl_glue::egl::EGLInitResult`

error: aborting due to 3 previous errors

Some errors have detailed explanations: E0063, E0308, E0599.
For more information about an error, try `rustc --explain E0063`.
error: Could not compile `libmlservo`.
```

**Comment by jdm** [comment]
You can test the simpleservo changes locally using `./mach build -d --libsimpleservo`.

**Comment by ferjm** [comment]
Yes, that worked on my laptop (OSX)

```
~/dev/mozilla/servo player-context*
❯ ./mach build -d --libsimpleservo
   Compiling simpleservo v0.0.1 (/Users/ferjm/dev/mozilla/servo/ports/libsimpleservo/api)
   Compiling simpleservo_capi v0.0.1 (/Users/ferjm/dev/mozilla/servo/ports/libsimpleservo/capi)
    Finished dev [unoptimized + debuginfo] target(s) in 2m 41s
Build Completed in 0:02:41
```

**Comment by CYBAI** [comment]
In `magicleap` build:

```
error[E0560]: struct `simpleservo::InitOptions` has no field named `gl_context_points`
   --> ports/libmlservo/src/lib.rs:160:9
    |
160 |         gl_context_points: Some(gl.gl_context),
    |         ^^^^^^^^^^^^^^^^^ help: a field with a similar name exists: `gl_context_pointer`

error[E0560]: struct `simpleservo::InitOptions` has no field named `display`
   --> ports/libmlservo/src/lib.rs:161:9
    |
161 |         display: Some(gl.display),
    |         ^^^^^^^ `simpleservo::InitOptions` does not have this field
    |
    = note: available fields are: `args`, `url`, `coordinates`, `density`, `vr_init` ... and 3 others

error: aborting due to 2 previous errors
```

**Pull request merged**